client/my-sites/site-settings/date-time-format/date-time-format-options.jsx

```diff
@@ -0,0 +1,337 @@
+import { phpToMomentDatetimeFormat } from 'lib/formatting';
+import wrapSettingsForm from '../wrap-settings-form';
+
+/**
+ * Module constants
+ */
+const defaultDateFormats = [ 'F j, Y', 'Y-m-d', 'm/d/Y', 'd/m/Y' ];
+const defaultTimeFormats = [ 'g:i a', 'g:i A', 'H:i' ];
+
+export class DateTimeFormatOptions extends Component {
+	state = {
+		customDateFormat: false,
+		customTimeFormat: false,
+		isLoadingSettings: true,
+	};
+
+	componentWillReceiveProps( nextProps ) {
+		const {
+			fields: {
+				date_format: dateFormat,
+				time_format: timeFormat,
+			},
+		} = nextProps;
+
+		if ( ! this.state.isLoadingSettings || '' === dateFormat || '' === timeFormat ) {
+			return;
+		}
+
+		this.setState( {
+			customDateFormat: ! includes( defaultDateFormats, dateFormat ),
+			customTimeFormat: ! includes( defaultTimeFormats, timeFormat ),
+			isLoadingSettings: false,
+		} );
+	}
+
+	setFormat = ( name, defaultFormats ) => event => {
+		const { value: format } = event.currentTarget;
+		this.props.updateFields( { [ `${ name }_format` ]: format } );
+		this.setState( {
+			[ `custom${ capitalize( name ) }Format` ]: ! includes( defaultFormats, format ),
+		} );
+	};
+
+	setDateFormat = this.setFormat( 'date', defaultDateFormats );
+
+	setTimeFormat = this.setFormat( 'time', defaultTimeFormats );
+
+	setCustomFormat = name => event => {
+		const { value: format } = event.currentTarget;
+		this.props.updateFields( { [ `${ name }_format` ]: format } );
+		this.setState( {
+			[ `custom${ capitalize( name ) }Format` ]: true,
+		} );
+	};
+
+	setCustomDateFormat = this.setCustomFormat( 'date' );
+
+	setCustomTimeFormat = this.setCustomFormat( 'time' );
+
+	/**
+	 * Adjust the current date and time to the site settings timezone.
+	 * The timezone can be formatted either as an UTC offset ("UTC+0"),
+	 * or as a timezone identifier ("Europe/London").
+	 * In the first case, the date-time must be adjusted by the UTC offset converted in minutes;
+	 * in the latter, it's enough to use the tz() method provided by Moment.js.
+	 *
+	 * @see http://momentjs.com/docs/#/manipulating/utc-offset/
+	 * @see http://momentjs.com/timezone/docs/#/using-timezones/parsing-in-zone/
+	 *
+	 * @return {Object} The timezone-adjusted Moment.js object of the current date and time.
+	 */
+	getNow = () => {
+		const {
+			fields: { timezone_string: timezoneString },
+			moment,
+		} = this.props;
+
+		return startsWith( timezoneString, 'UTC' )
+			? moment().utcOffset( timezoneString.substring( 3 ) * 60 )
+			: moment.tz( timezoneString );
+	}
+
+	dateFormatOption() {
+		const {
+			fields: { date_format: dateFormat },
+			isRequestingSettings,
+			translate,
+		} = this.props;
+		const { customDateFormat: isCustomFormat } = this.state;
+
+		const now = this.getNow();
+
+		return (
+			<FormFieldset>
+				<FormLabel>
+					{ translate( 'Date Format' ) }
+				</FormLabel>
+				{ defaultDateFormats.map( ( format, key ) =>
+					<FormLabel key={ key }>
```

Comment:
be careful of using loop indices as keys. it may work in this case, but it's really not a `key` - more an `index`. I would recommend either retaining the `index` name or using an actual unique id as the key, such as `format` itself